Add support for MSI AM5/LGA1851 Motherboards using the NCT6687D SIO Controller

Alcolawl opened the pull request:
### Background
It appears that new MSI Motherboards with AMD Socket AM5 or Intel Socket LGA1851 that utilize the Nuvoton NCT6687D Super I/O Controller behave differently than motherboards of the past with the same model controller. This led to issues with sensors being incorrectly detected, mainly fan speed and duty cycle readouts as well as the ability to control the system fan headers. After an unbelievable amount of work, including hours of trial and error, community testing, etc., I believe it's time we merge this into LibreHardwareMonitor.

### Motherboard Support
This PR includes support for the following 15 MSI Motherboards:

1. B850 GAMING PLUS WIFI
2. PRO B850-P WIFI 
3. MAG B850 TOMAHAWK MAX WIFI
4. X870 GAMING PLUS WIFI
5. MAG X870 TOMAHAWK WIFI
6. PRO X870-P WIFI
7. MPG X870E CARBON WIFI
8. MPG X870E EDGE TI WIFI
9. MAG X870E TOMAHAWK WIFI
10. MEG X870E GODLIKE
11. MEG Z890 ACE
12. MPG Z890 CARBON WIFI
13. MPG Z890 EDGE TI WIFI
14. MAG Z890 TOMAHAWK WIFI
15. PRO Z890-P WIFI

This includes temperature and voltage sensor layouts, readouts tested and vetted by owners of the actual motherboards, as well as confirmed control of system fans connected to those systems.

### Changes Required and Thought Process behind it

**New Super I/O Controller - The NCT6687D-R**
Because motherboards prior to this generation also use the Nuvoton NCT6687D SIO Controller, the main concern was obviously maintaining legacy compatibility while expanding upon it to support these new motherboards. Since the controllers on new and old motherboards are identified as the same model but obviously behave quite differently, the solution was to create a new chip named the NCT6687D-R (Which is an actual model, it just identifies itself with the same register values as NCT6687D). All 15 of the supported motherboards above are explicitly assigned the new chip. This allows us to write code strictly for the NCT6687DR while maintaining the known good working code for the NCT6687D. Unfortunately, this also means that future support for the ever-growing library of MSI motherboards will need to be added individually, but it's the best workaround I could come up with.

**New Fan Control Logic**
The main focus of this project was to restore control of the system fan headers for use in third party programs like FanControl. The issue extends from the fan hardware registers being stored in different places and the fan control logic being changed. According to correspondence with Nuvoton, MSI has modified the NCT6687D, which MSI refuses to share documentation on. Each fan detected by the Super I/O controller has a dedicated 7-Point Curve it obeys. The solution for restoring control was to create a special function that is only called during a "SetControl()" call (in NCT677X.cs) when chip == NCT6687DR. This function will set the all points along the aforementioned 7-point curve to the same value so the fan has no choice but to obey. While the fans may react slower than an end user may desire, they do react to user input. This is something we've focused on for quite a while to try to fix, but with limited resources and no vendor support, we've hit a brick wall.

### Wrapping Up
Countless issues have been opened in this repo asking to add support for this SIO controller or new MSI motherboards, the main one being #1513, which grew into an incredible size. Hence why I created a fork to have a separate place to tinker, discuss, and add support for more and more motherboard models.

Please let me know what you think, what changes are required, tweaks, etc. and thank you in advance for your time!

xSalts commented:
Would love to see this in the release version!

GaXWork commented:
I think the newer Asus MB also need a update to get working, not sure its the same changes as MSI

danielrweiss commented:
I would also love to see this merged in. This has resolved all issues I was having with Libre and FanControl issues reading sensors and controlling fan speeds on a system with an MSI MAG X870E Tomahawk motherboard.

PhyxionNL commented on LibreHardwareMonitorLib/Hardware/Motherboard/Lpc/Nct677X.cs:
```diff
@@ -409,7 +478,13 @@ public void SetControl(int index, byte? value)
                 Thread.Sleep(50);
 
-                WriteByte(FAN_PWM_COMMAND_REG[index], value.Value);
+                if (Chip is Chip.NCT6687DR){ // For MSI AM5/LGA1851 NCT6687D functionality
```
Just going to comment once, can you reformat all documents to follow code style? For example, `){` should be `) {`, same with `else{` to `else {`

Alcolawl replied:
Done deal. Reformatted the code to follow the style and tested to ensure functionality remained in-tact.

PhyxionNL commented:
Thanks @Alcolawl!

chrisgleissner commented:
Thanks a lot @demomanca and @Alcolawl for the amazing research you did to get fan control to work on the new MSI boards. As of yesterday, the official FanControl release now includes your work via the latest LHM main branch build. Would not have happened without your efforts.

Ricardaso commented:
Hello,

I am new to Libre, but I have Z890 Gaming Plus WiFi board, CPU and Pump fans can me found and controlled, but SYSFAN sensors cannot be found in FanControl. Though HWinfo can see the fans, also SYSFANS can be controlled in MSI center.
MSI Z890 GAMING PLUS WIFI (MS-7E34) (Nuvoton NCT6687D)

Absolution13 commented:
Hello, 

any updated for the Z890s Mobos? On MSI Z890 Gaming Plus WiFi the SYS Fans are still not detected.

Alcolawl commented:
> Hello,
> 
> I am new to Libre, but I have Z890 Gaming Plus WiFi board, CPU and Pump fans can me found and controlled, but SYSFAN sensors cannot be found in FanControl. Though HWinfo can see the fans, also SYSFANS can be controlled in MSI center. MSI Z890 GAMING PLUS WIFI (MS-7E34) (Nuvoton NCT6687D)

> Hello,
> 
> any updates for the Z890s Mobos? On MSI Z890 Gaming Plus WiFi the SYS Fans are still not detected.

I will look into it later this week or this weekend. Lots of stuff in real life driving me crazy right now. I also honestly did not expect to keep maintaining / adding motherboards but I'll see if I can get a PR going to get the MSI Z890 Gaming Plus WiFi added. If not, I'll link a build in my personal fork of LHM that should work for you guys.

Alcolawl commented:
@Absolution13 @Ricardaso 

Give this build a shot and report back! **Please read the README** for installation and usage instructions.
[LHM_NCT6687DR_Build2571.zip](https://github.com/user-attachments/files/21196894/LHM_NCT6687DR_Build2571.zip)

Absolution13 commented:
@Alcolawl I reinstallted the 4.8Net Version just to be sure and copied the new LHM files you provided. FC only detects the GPU fans now.

Absolution13 commented:
> > @Alcolawl I reinstalled the 4.8Net Version just to be sure and copied the new LHM files you provided. FC only detects the GPU fans now.
> 
> Try downloading the 4.8 Net version of LHM and replacing the files form my build and see if the fans and everything get detected properly.

After using the LHM files from the default 4.8Net version the CPU and Pump Fans are detected again, still no detection for any SYS Fan.

Absolution13 commented:
> > > > @Alcolawl I reinstalled the 4.8Net Version just to be sure and copied the new LHM files you provided. FC only detects the GPU fans now.
> > > 
> > > 
> > > Try downloading the 4.8 Net version of LHM and replacing the files form my build and see if the fans and everything get detected properly.
> > 
> > 
> > After using the LHM files from the default 4.8Net version the CPU and Pump Fans are detected again, still no detection for any SYS Fan.
> 
> Apologies, I did not realize LHM has changed since the last time I worked with it. I was able to replicate your issue on my end. The Zip I provided was missing some extra DLLs. Try this one.
> 
> [LHM_NCT6687DR_Build2572.zip](https://github.com/user-attachments/files/21197360/LHM_NCT6687DR_Build2572.zip)

With the 2572 Build FC is going back to detecting only the GPU fans again. :(

Alcolawl commented:
Please download LHM, run it and take a screenshot, then replace the files in there from my build and take another screenshot and reply with both.

Absolution13 commented:
> Please download LHM, run it and take a screenshot, then replace the files in there from my build and take another screenshot and reply with both.

Default LHM from the 4.8Net Version: 
<img width="1946" height="1059" alt="image" src="https://github.com/user-attachments/assets/15fd3aa2-5073-4e03-951f-3c55813119fd" />

With the Files from the 2572 Build: 

<img width="1933" height="1025" alt="image" src="https://github.com/user-attachments/assets/9531bd34-bb07-403c-b9fe-ba31607aa8c4" />

Alcolawl commented:
> > Please download LHM, run it and take a screenshot, then replace the files in there from my build and take another screenshot and reply with both.
> 
> Default LHM from the 4.8Net Version: 
> <img width="1946" height="1059" alt="image" src="https://github.com/user-attachments/assets/15fd3aa2-5073-4e03-951f-3c55813119fd" />
> 
> With the Files from the 2572 Build: 
> 
> <img width="1933" height="1025" alt="image" src="https://github.com/user-attachments/assets/9531bd34-bb07-403c-b9fe-ba31607aa8c4" />
> 
> 

Nono, not Fan Control. Download and run the - latest LibreHardwareMonitor release itself, screenshot, then replace the files with the files from my build, run LHM again and take another screenshot.

Alcolawl commented:
@Absolution13 Other users posting something similar. It's not detecting your motherboard properly. Mind attaching a report from LibreHardwareMonitor? Just open LHM and click File > Save Report, then attach it here. If you're concerned about privacy, you can remove everything from the CPUID section down in notepad or your favorite code editor.

Absolution13 commented:
> @Absolution13 Other users posting something similar. It's not detecting your motherboard properly. Mind attaching a report from LibreHardwareMonitor? Just open LHM and click File > Save Report, then attach it here. If you're concerned about privacy, you can remove everything from the CPUID section down in notepad or your favorite code editor.

No worries here it is 
[LibreHardwareMonitor.Report.txt](https://github.com/user-attachments/files/21201693/LibreHardwareMonitor.Report.txt)

Alcolawl commented:
I've triple-checked the code. I don't understand what's going on here. The general workflow for identifying the Motherboard, SIO Controller, etc goes like this:

1. LHM grabs the name of your motherboard.
2. Now that we have the name, we detect what chip ID you have (The ID returned by the Nuvoton controller on your motherboard. In this case is 0xD592, where D5 is the Chip and 92 is the revision).
3. **IF** Chip 0xD5 and revision 0x92 is returned, and the motherboard model matches your motherboard (model.Z890_GAMING_PLUS_WIFI), then force the chip = chip.NCT6687DR.
4. So, chip.NCT6687DR is true. From there, conditional statements take care of the rest where NCT6687DR is treated as it's own unique chip with it's own registers, logic, etc. 

This has worked with 15+ motherboards, but now it suddenly isn't with 2 new ones. The thing is, you can follow along with this, even if what I said above might seem like jibberish. The report you attached clearly shows the information the code is looking for is correct. Everything in quotes is straight from your LHM report:

"MSI Z890 GAMING PLUS WIFI (MS-7E34)"
`            case var _ when name.Equals("Z890 GAMING PLUS WIFI (MS-7E34)", StringComparison.OrdinalIgnoreCase):
                return Model.Z890_GAMING_PLUS_WIFI;`

That checks out. So your motherboard is detected internally as Z890_GAMING_PLUS_WIFI. Now that we have that, we detect the chip using the ID and revision:

"LPC Nct677X

Chip Id: 0xD592
Chip Revision: 0x92"

```
            case 0xD5:
                switch (revision)
                {
                    case 0x92:                                
                        switch (motherboard.Model)
                        {
                            case Model.B840P_PRO_WIFI:
                            case Model.B850_GAMING_PLUS_WIFI:
                            case Model.B850P_PRO_WIFI:
                            case Model.B850M_MORTAR_WIFI:
                            case Model.B850_TOMAHAWK_MAX_WIFI:
                            case Model.B850_EDGE_TI_WIFI:
                            case Model.X870_GAMING_PLUS_WIFI:
                            case Model.X870_TOMAHAWK_WIFI:
                            case Model.X870P_PRO_WIFI:
                            case Model.X870E_TOMAHAWK_WIFI:
                            case Model.X870E_CARBON_WIFI:
                            case Model.X870E_EDGE_TI_WIFI:
                            case Model.X870E_GODLIKE:
                            case Model.Z890_ACE:
                            case Model.Z890_CARBON_WIFI:
                            case Model.Z890_TOMAHAWK_WIFI:
                            case Model.Z890_EDGE_TI_WIFI:
                            case Model.Z890P_PRO_WIFI:
                            case Model.Z890A_PRO_WIFI:
                            case Model.Z890_GAMING_PLUS_WIFI:
                            case Model.Z890S_PRO_WIFI_PROJECT_ZERO:
                                chip = Chip.NCT6687DR; // MSI AM5/LGA1851 Compatibility
                                break;
                            default:
                                chip = Chip.NCT6687D;
                                break;
                        }
                        logicalDeviceNumber = WINBOND_NUVOTON_HARDWARE_MONITOR_LDN;
                        break;
                }
```

That checks out. Chip ID is 0xD592, just like all the others, then we check if your motherboard matches the list in the code above, which it does. Again, it's looking for Model.Z890_GAMING_PLUS_WIFI, which is there, clear as day. 

Unfortunately, that's kind of where things fall apart. It's not doing what the code says. I've written code professionally for 5+ years, and on the side as a hobby for 15. The funny thing is that most of the time, computers do exactly what we tell them to. That is to say, that it's usually the programmer's fault when they get a result they don't expect. However, in this case, I've simply added 2 lines to preexisting code that I wrote myself and I know works for other motherboards. The only thing I can think of at this point is that my LibreHardwareMonitor binary is not building properly. Things have apparently changed since earlier this year. I've scoured the terminal for errors and things that would stand out as problematic, but nothing jumps out. I've tested every build I've posted here on my end to make sure my motherboard still works, just as the logic above intends, and it works perfectly. For whatever reason, it refuses to detect your motherboard (as well as another, different motherboard I tried adding for another user). 

You can try this fresh build where I legitimately re-cloned the LHM git and added my changes again to see if something was wrong on my PC. Replace the files in LHM, not FanControl and try again? Fingers crossed, but I'm stumped right now.

[LHM_NCT6687DR_Build2573.zip](https://github.com/user-attachments/files/21205435/LHM_NCT6687DR_Build2573.zip)

Alcolawl commented:
@Absolution13 there we go! Yeah the files I provide are Plug n Play, you don't need to do any compilation. 

Can you try plunking those same files down in FanControl's dir and giving that a shot? You will most likely have to manually pair speed sensors with the controls and re-make a config. Keep that in mind.

Absolution13 commented:
> @Absolution13 there we go! Yeah the files I provide are Plug n Play, you don't need to do any compilation.
> 
> Can you try plunking those same files down in FanControl's dir and giving that a shot? You will most likely have to manually pair speed sensors with the controls and re-make a config. Keep that in mind.

Fan Control identifies the fans and their current RPMs but it is not sensitive to adjustments. :(

Absolution13 commented:
> > > @Absolution13 there we go! Yeah the files I provide are Plug n Play, you don't need to do any compilation.
> > > Can you try plunking those same files down in FanControl's dir and giving that a shot? You will most likely have to manually pair speed sensors with the controls and re-make a config. Keep that in mind.
> > 
> > 
> > Fan Control identifies the fans and their current RPMs but it is not sensitive to adjustments. :(
> 
> The fans should respond, just incredibly slowly. Like 1-2% per second. Make sure to calibrate the controls, you may have to manually tweak the values.
> 
> Unfortunately, MSI's implementation of system fan control via that controller is really slow to respond and we haven't found a way to get them to respond like the CPU and Pump headers.

Gotcha ill try to calibrate it more precisely manually, thanks alot for your help!!

Ricardaso commented:
Mine works fine and ramps quite quick :) I also pasted content from last ZIP to my fan control dir.

Alcolawl commented:
Awesome! Thanks for the help during this process, fellas.

Shnig20 commented:
Hey Alcolawl, great work. Fair play! Just a small thing. On my x870 Gaming Plus WiFi the fan I have connected to the EZ Conn Connector does not show up (no RGB, just a fan).

Let me know what details you want/need. Thanks.

Shnig20 commented:
> Hey Alcolawl, great work. Fair play! Just a small thing. On my x870 Gaming Plus WiFi the fan I have connected to the EZ Conn Connector does not show up (no RGB, just a fan).
> 
> Let me know what details you want/need. Thanks.

Just a bit more added info. The "EZ Conn" uses the JAF_1 header on the motherboard. I have 1 P14 max attached to it.

Please see attached for more info. It shows up in HWInfo not LHM.
<img width="1851" height="1028" alt="Screen Shot for EZ Conn 2" src="https://github.com/user-attachments/assets/c1d35eb4-dfe8-4526-8db7-38f7341551be" />
<img width="1848" height="1032" alt="Screen Shot for EZ Conn" src="https://github.com/user-attachments/assets/e40e3dea-6706-4531-a374-86c5f17c8bae" />

Alcolawl commented:
> > Hey Alcolawl, great work. Fair play! Just a small thing. On my x870 Gaming Plus WiFi the fan I have connected to the EZ Conn Connector does not show up (no RGB, just a fan).
> > Let me know what details you want/need. Thanks.
> 
> Just a bit more added info. The "EZ Conn" uses the JAF_1 header on the motherboard. I have 1 P14 max attached to it.
> 
> Please see attached for more info. It shows up in HWInfo not LHM.
> <img alt="Screen Shot for EZ Conn 2" width="1851" height="1028" src="https://private-user-images.githubusercontent.com/221110968/467032004-c1d35eb4-dfe8-4526-8db7-38f7341551be.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTI2Nzk3OTEsIm5iZiI6MTc1MjY3OTQ5MSwicGF0aCI6Ii8yMjExMTA5NjgvNDY3MDMyMDA0LWMxZDM1ZWI0LWRmZTgtNDUyNi04ZGI3LTM4ZjczNDE1NTFiZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNzE2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDcxNlQxNTI0NTFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yYWMwMzdlZTRhMzg5ZTgxYzQ1ZmI0ZTkzMmJmMWZiZWYwNTg3ZmZhM2JkZWMwYjQwN2E0MGFiY2ExODA3ZmQ3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.Zn-BoSWZ37NgfGsYi4JUB8wqWIIHaT0lj-OPcorjCQU"> <img alt="Screen Shot for EZ Conn" width="1848" height="1032" src="https://private-user-images.githubusercontent.com/221110968/467032024-e40e3dea-6706-4531-a374-86c5f17c8bae.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTI2Nzk3OTEsIm5iZiI6MTc1MjY3OTQ5MSwicGF0aCI6Ii8yMjExMTA5NjgvNDY3MDMyMDI0LWU0MGUzZGVhLTY3MDYtNDUzMS1hMzc0LTg2YzVmMTdjOGJhZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNzE2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDcxNlQxNTI0NTFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jZTZiNTc1OWNkOTNiZWU2NzBmOWM5Zjc3ZmFlMWRhMTk3MGYyM2ZjMDk1YTk4ZGIwNjUyMWQxZTVlYmY5MGM1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.JD9c4QybpB8x7mzsH0HdDxBCLz03TRBjxP2H5V1TsS8">
> 
> As you can see I also tried adding the sensor via the HWInfo pluggin but no luck and FYI I imported the latest ZIP in this discussion to both my Fan Control folder and my LHM folder both of which are running the Net 4.X version.

I can give it a shot but I have zero experience and zero way of testing the EZ Conn port on my end.

[Please open an issue on my fork](https://github.com/Alcolawl/LibreHardwareMonitor/issues), so as not to clutter this issue, that's closed, and we'll see what we can do. Include all the information from this comment as well as a report from LibreHardwareMonitor. HWInfo doesn't help me because it's closed source. With no other monitor programs running, download and run the latest version of LibreHardwareMonitor, go to File > Save Report, and attach it on the issue you open. That should give me the register readout I need to try to pin-point the EZ-Connect registers and write the corresponding code.

jackjojo21-svg commented:
Hello

I'm a user of the MSI Z890 Gaming WiFi motherboard, and unfortunately, fan detection still isn't working.

I'm still using the LHM from July 13th to get my fans recognized.

I also just tried the one from https://github.com/Alcolawl/LibreHardwareMonitor:
LHM_NCT6687DR_Build25101.zip

but unfortunately, it didn't work.
<img width="928" height="1015" alt="Screenshot 2025-10-30 105623" src="https://github.com/user-attachments/assets/f7a68ba8-5388-4a44-a7d5-b85d71a90d5d" />

Alcolawl commented:
You have to update to the latest release of FanControl (.NET 4.8 Version) to use the latest release of my LHM fork. Additionally, you'll have to completely remake your configuration, as well as install the PawnIO driver. All of this is included in the README that comes with my LHM releases.

Alcolawl commented:
> @Alcolawl Thank you for the response. I tried your suggestion and set all the ramp up/down speeds in BIOS to .01s, but I do not see any difference in the actual fan ramp up/down speed. My point here is that even MSI's own BIOS fan curve, and the fan control tool shipped with MSI Center, have the exact same problem, which leads me to believe MSI's fan controller configuration is broken for the SYS_FAN headers, and this is something that likely cannot be fixed through software, and MSI needs to fix the underlying firmware issue (assuming that this is a firmware issue and not a hardware issue). Is there any evidence that this is not the case? Have you came across any case where someone is using an MSI X870/X870 motherboard but is NOT experiencing slow fan ramp up/down?

No one that I've talked to with these motherboards has indicated the fan system fan behavior has been anything different than what you're experiencing. 

The Nuvoton documentation is pretty extensive and details all of the different ways that fan behavior can be governed. It's actually way more complex than you might imagine. Instead of jumping to the conclusion that their firmware is broken, my guess is that they deliberately set this behavior (lord knows why) and it can be controlled or changed by a certain hardware register, but until it's found, there's no way of proving that theory out. The Regular Nuvoton NCT6687D has hardware registers to set the fan control behavior multiple different ways, but when we tried to adjust them on this MSI flavored chip, they didn't work. However, this doesn't mean it's broken, MSI moved around, added, changed dozens of registers. So my hope is that it's right under our noses, we just need to be nudged to the correct one.
